hoomd/md/tune/nlist_buffer.py

```diff
@@ -0,0 +1,281 @@
+
+    def get_tps(self):
+        if self._last_walltime is None or self._last_timestep is None:
+            return None
+        delta_w = self._simulation.walltime - self._last_walltime
+        delta_t = self._simulation.timestep - self._last_timestep
+        # We divide by 1_000 to reduce the gradient size for optimization.
+        return delta_t / (_SCALE_TPS * delta_w)
+
+
+class _NeighborListBufferInternal(hoomd.custom._InternalAction):
+    _skip_for_equality = {"_simulation", "_tunable"}
+
+    def __init__(
+        self,
+        nlist: NeighborList,
+        solver: hoomd.tune.solve.Optimizer,
+        maximum_buffer: float,
+    ):
+        self._simulation = None
+        self._tuned = False
+        param_dict = hoomd.data.parameterdicts.ParameterDict(
+            nlist=SetOnce(NeighborList),
+            solver=SetOnce(hoomd.tune.solve.Optimizer),
+            maximum_buffer=OnlyTypes(float,
+                                     postprocess=self._maximum_buffer_post))
+        param_dict.update({
+            "nlist": nlist,
+            "solver": solver,
+            "maximum_buffer": maximum_buffer
+        })
+        self._param_dict.update(param_dict)
+
+        # Setup default log values
+        self._last_tps = 0.0
+        self._best_buffer_size = nlist.buffer
+        self._max_tps = 0.0
+
+    def act(self, timestep: int):
+        tps = self._tunable.y
+        if tps is not None:
+            self._last_tps = tps * _SCALE_TPS
+        if not self.tuned:
+            if tps is not None and tps > self._max_tps:
+                self._best_buffer_size = self._tunable.x
+                self._max_tps = self._last_tps
+            self._tuned = self.solver.solve([self._tunable])
+
+    def _maximum_buffer_post(self, value: float):
+        if self._simulation is not None:
+            self._tunable.domain = (0, value)
+        return value
+
+    def attach(self, simulation):
+        self._simulation = simulation
+        self._tunable = self._make_tunable(self.nlist)
+
+    def _make_tunable(self, nlist):
+        return hoomd.tune.ManualTuneDefinition(
+            get_y=_IntervalTPS(self._simulation),
+            target=0.0,
+            get_x=self._get_buffer,
+            set_x=self._set_buffer,
+            domain=(0, self.maximum_buffer),
+        )
+
+    def _get_buffer(self):
+        return self.nlist.buffer
+
+    def _set_buffer(self, new_buffer):
+        self.nlist.buffer = new_buffer
+
+    def detach(self):
+        self._simulation = None
+        self._tunable = None
+
+    @property
+    def tuned(self):
+        """bool: Whether the neighbor list buffer is considered tuned.
+
+        Like other tuners, the tuner is considered tune when the specified
+        solver returns ``True`` when solving. See `hoomd.tune` for more
+        information on tuning criteria.
+        """
+        return self._tuned
+
+    @hoomd.logging.log
+    def max_tps(self):
```

Comment:
`max_tps` doesn't update after the first time it is set (see example output in review comment).